### DISCUSSION TITLE
Vue3's recent update causing "dirty computed getter" warning due to some inner workings of useStore() with `mapState()`

### DISCUSSION THREAD

**Discussion by ayochim** [vuejs/pinia#2585] [Help and Questions] [answered]
Paraphrase: when a readonly computed getter contains useStore(), vue3 throws a warning about self-triggering computed getters.
See the following for more information and discussion.
https://github.com/vuejs/core/issues/10341
https://github.com/vuejs/core/discussions/10340
https://github.com/vuejs/core/pull/10299
https://github.com/vuejs/core/commit/f7ba97f9754a9882c1f6b1c07ca1a4040479dd13
[playground reproduction](https://play.pinia.vuejs.org/#eNqNVMFu2zAM/RXBFydAIrVIV2BZGmQbCmw7bMM67DLvoNpMok6WBElOAwT+91Fy7DhZg+YkkXwkHymKu+S9MXRTQTJNZh5KI7mHeaYImZn5J7BAhCN+DcR5jcKGywoIn5Ldjtf1jJnXoCpA1YtQ3kAiQNuSy1/ctsAZ67gQlFxuhfFEcrW6yxLvsgQxojTaerIjJTcPHqGkJkurS5IaoQRP3/UglYOHyKqFUJbrSnmw1Lu0ByxgKRR81CgrUL7DY4dCQNhGHKJ4JePZRw92ocSCez4YxishFnxlFdlLhHSVTkm6Bik1edZWFumoAdThqKOQY9TKQ4H9aWyU0rbSQVvPiPxOeToiqUr/DKNbcK6HSHXGmq7Nk1HSlDdGd/rktMK3jjGzvQH72WXJEiw1yFmy9t64KWN5odCtACk2lirwTJmSLRDGLLZQlDAudLmY0Bt6/ZYVwvm+noIrx49WPzvs9RNm2lca8zBUbsCOLagCp8JemvfE7Sj3ie18/jglTUYW7+ehMWkBG6+1dGNuxKVE+z6LW/qGXjMpHhkmYgL5bf/ryLiA8pLoLXRxRa9v6G2Mupnsg5YhahwGnAV8fu9yrZZidfL4YcKEBPvNeKHV8RBwHM3nL1HnbQUdxXwN+d8X9E9u27D+biG+QK8sz+0KfGO+f/gKW7x3xlIXldwP3BnjD3BaVoFjA/tQqQJp93CR7ec4y0Ktfrr7rQfl2qIC0cO/im0OX/Zc6Qe6EzrpdfGwMLCFJwvjeLk0+6e3VSwsR92HPtopmeqWCr6R84dNddcPPUjjUsWfPhiSu3lDtnFQiMT4g6vhQcf3upuwCIK6XUOIHvFY0DCp/wFs/gN7)

here's the warning:
[Vue warn] Computed is still dirty after getter evaluation, likely because a computed is mutating its own dependency in its getter. State mutations in computed getters should be avoided.  Check the docs for more details: https://vuejs.org/guide/essentials/computed.html#getters-should-be-side-effect-free

**Comment by ghost** [comment]
Hello, I am providing this example, where the warning is being generated.
Previously, I was able to use Pinia stores inside Vue 3 computed properties as follows:

```javascript
export const useExampleStore = defineStore({
    id: "example",
    state: () => ({
        foo: "foo value"
    }),
    getters: {
        fooBar: () => `${foo} bar`
    },
});

const val = ref("");
const myFooComputed = computed(() => useExampleStore().fooBar);
```
However, with the latest version, I’m receiving the following warning:

* *_[Vue warn] Computed is still dirty after getter evaluation, likely because a computed is mutating its own dependency in its getter._*

This seems to suggest that there’s an issue with how computed properties are being handled in the latest version.

I believe this is a significant issue that needs to be addressed. Any guidance or updates on this would be greatly appreciated.

**Reply by Doctor-wu** [reply, marked as answer]
In your case, it can work in this way.
```typescript
const exampleStore = useExampleStore();
const foll = computed(() => exampleStore.fooBar)
```

**Reply by ghost** [reply]
That is a good solution, but when the project is large, and I need only one property of the store, I usually do like this:
```javascript
// const exampleStore = useExampleStore();
const foll = computed(() => {
    const exampleStore = useExampleStore();
    return exampleStore.fooBar;
})
 ```
am I wrong?, thanks and best regards

**Reply by Doctor-wu** [reply]
usexxxStore in pinia has side-effect, and it's not suggested to invoke in the computed after vue3.4 since it will cause unexpected track.

**Reply by posva** [reply]
@Doctor-wu is right, you should put the store outside of the computed property

**Reply by Doctor-wu** [reply]
@posva Thanks for repling ❤️, there have another problem is `mapState` will return a series of computed and will call `useStore` inside that computed. That will cause warning as well, are there any approaches to avoid this?

**Reply by posva** [reply]
Yeah, `mapState()` is a bit special... It was fixed and apparently the issue is back again: https://github.com/vuejs/pinia/issues/2551

**Reply by Doctor-wu** [reply]
Glad to know this is a known issue, I will attempt to fix this (maybe I just can't🤣)

**Reply by sadovsf** [reply]
This solution cannot be applied universaly. When project grows you quickly run into situations where in some function you need to get something from store and coincidently it will get called somewhere down the line through computed with no way of knowing it in advance or no reasonable place where you could pre-cache store. We have for example model classes which have many getters for translation from raw reactive data object inside to actual other models. Imagine translating SQL ID to actual class instance as part of given getter. There is no reasonable way how to know who will call that getter and although it just does read operations it currently warns (hopefully not break) on this issue as it needs to get into store to do this translation.

**Comment by p1xel007** [comment]
I'm having the same issue

**Comment by sOMarkus** [comment]
We have same issue using Options API for modifiable state
[Modifiable state](https://pinia.vuejs.org/core-concepts/state.html#Modifiable-state)

I've already updated to Vue 3.4.19 and Pinia 2.1.7 - but still same issue.

**Usage in Component:**
```js
computed: {
    ...mapWritableState(useTemporaryStore, ["currentObject"]),
  },
``` 

**Store:**
```js
import { defineStore } from "pinia";

export const useTemporaryStore = defineStore("TemporaryStore", {
  state: () => {
    return {
      currentObject: null,
    };
  },
});
```

**Comment by bitbybit** [comment]
I am getting this warning when getter is accessing another getter from a different store: https://pinia.vuejs.org/core-concepts/getters.html#Accessing-other-stores-getters

**Comment by gabrielwillemann** [comment]
I have the same issue using `mapWritableState`

**Reply by sOMarkus** [reply]
Version 3.4.20 fixed the warnings with mapWritableState
[see discussion here](https://github.com/vuejs/core/issues/10341#event-11912914375)

And [Commit](https://github.com/OnlyWick/core/commit/23953dd795c71ea074f348a775d2242314d8ea99)

`dx(reactivity): disable recursive computed warning by default`
`Now can be enabled with app.config.warnRecursiveComputed option.`
